src/app/AttributeAccessInterface.h

```diff
@@ -41,23 +42,143 @@
+ *
+ * AttributeReportBuilder builder;
+ * ReturnErrorOnFailure(builder.PrepareAttribute(...));
+ * ReturnErrorOnFailure(builder.Encode(...));
+ * ReturnErrorOnFailure(builder.FinishAttribute());
+ */
+class AttributeReportBuilder
 {
 public:
-    AttributeValueEncoder(TLV::TLVWriter * aWriter, FabricIndex aAccessingFabricIndex) :
-        TagBoundEncoder(aWriter, TLV::ContextTag(to_underlying(AttributeDataIB::Tag::kData))),
-        mAccessingFabricIndex(aAccessingFabricIndex)
+    /**
+     * PrepareAttribute encodes the "header" part of an attribute report including the path and data version.
+     * Path will be encoded according to section 10.5.4.3.1 in the spec.
+     * Note: Only append is supported currently (encode a null list index), other operations won't encode a list index in the
+     * attribute path field.
+     */
+    CHIP_ERROR PrepareAttribute(AttributeReportIBs::Builder & aAttributeReportIBs, const ConcreteDataAttributePath & aPath,
+                                DataVersion aDataVersion);
+
+    /**
+     * FinishAttribute encodes the "footer" part of an attribute report (it closes the containers opened in PrepareAttribute)
+     */
+    CHIP_ERROR FinishAttribute();
+
+    template <typename... Ts>
+    CHIP_ERROR Encode(Ts... aArgs)
+    {
+        return DataModel::Encode(*mAttributeDataIBBuilder.GetWriter(), TLV::ContextTag(to_underlying(AttributeDataIB::Tag::kData)),
+                                 std::forward<Ts>(aArgs)...);
+    }
+
+private:
+    AttributeReportIB::Builder mAttributeReportIBBuilder;
+    AttributeDataIB::Builder mAttributeDataIBBuilder;
+};
+
+/**
+ * The AttributeValueEncoder is a helper class for filling report payloads into AttributeReportIBs.
+ * The attribute value encoder can be initialized with a AttributeEncodeState for saving and recovering its state between encode
+ * sessions (chunkings).
+ *
+ * When Encode returns recoverable errors (e.g. CHIP_ERROR_NO_MEMORY) the state can be used to initialize the AttributeValueEncoder
+ * for future use on the same attribute path.
+ */
+class AttributeValueEncoder
+{
+public:
+    class ListEncodeHelper
+    {
+    public:
+        ListEncodeHelper(AttributeValueEncoder & encoder) : mAttributeValueEncoder(&encoder) {}
+
+        template <typename... Ts>
+        CHIP_ERROR Encode(Ts... aArgs) const
+        {
+            return mAttributeValueEncoder->EncodeListItem(std::forward<Ts>(aArgs)...);
+        }
+
+    private:
+        AttributeValueEncoder * mAttributeValueEncoder;
+    };
+
+    class AttributeEncodeState
+    {
+    public:
+        AttributeEncodeState() : mAllowPartialData(false), mCurrentEncodingListIndex(kInvalidListIndex) {}
+        bool AllowPartialData() const { return mAllowPartialData; }
+
+    private:
+        friend class AttributeValueEncoder;
+
+        bool mAllowPartialData              = false;
+        ListIndex mCurrentEncodingListIndex = kInvalidListIndex;
```

Comment:
What do these members mean?  They need to be documented. That includes documenting what valid vs invalid list index means for the mCurrentEncodingListIndex and documenting what mAllowPartialData means.  Allow what sort of partial data where?